### DISCUSSION TITLE
Update process

### DISCUSSION THREAD

**Discussion by SchraderR** [SchraderR/sp-tarkov-client#130] [Ideas]
# Update Process
_I will try to write down my ideas to make them more transparent._

# Bad Workflow _(we are here)_
_Create Flowchart_
_Start App → Scan Mods  → Map Mod Names with ext. List → `Use Informations to check for update`_

### Problem
There is no link of installed mods to the files and information on the hub. Therefore, a temporary mapping (potential issues and more) is needed to create this linking, see config https://github.com/SchraderR/sp-tarkov-client-configuration/blob/master/config.json;

# Good Workflow
_Create Flowchart_
_Start App → Scan Mods → `Use Informations to check for update`_

### Idea
Mod Authors should be responsible and add additional information to create this link. The best or fastest solution would be to add the SPT-Hub file ID in the `package.json` or `DLL` file.
One Problem could be of re-uploaded mods not edited correctly and ship old hub file ID with the new files. _(Not my problem)_

Example Hub-ID: `https://hub.sp-tarkov.com/files/file/1702-sp-eft-manager/`
`1702 `— Hub ID, not changeable
`sp-eft-manager` —  name, changeable

**Note: With the Hub ID, information can be fetched over an official API or with web scrapping**

#### Example `package.json`
```
{
  "name": "Mod Name",
  "version": "x.x.x",
  "hubId": 1702
}
```
#### Example DLL (*.csproj)
```
<ItemGroup>
  <AssemblyAttribute Include="System.Reflection.AssemblyMetadataAttribute">
    <_Parameter1>HubId</_Parameter1>
    <_Parameter2>1702</_Parameter2>
  </AssemblyAttribute>
</ItemGroup>
```

Output
`[assembly: AssemblyMetadata("HubId", "1702")]`
Reference: https://learn.microsoft.com/en-us/dotnet/standard/assembly/set-attributes-project-file#set-arbitrary-attributes

**Comment by maxloo2** [comment]
I think the "bad workflow" is not as bad as you think, asking the modders to add the SPT Hub ID to their mod is more or less impossible without the help from SPT staffs. They will have somehow enforce this, which I don't think is worth the while for anyone...

Anyway, the update flow: mod list has to be saved locally, we can refernece this list to find the currently installed version of all the mods, this will not work for already installed mods unless we can somehow recognize the mods from their files (e.g. package.json), but the issue is that some mods are just .DLLs, some are split between client & server...

**Reply by maxloo2** [reply]
![image](https://github.com/SchraderR/sp-tarkov-client/assets/4709356/bde83d13-c507-4a16-bf50-b6993110b94a)
Slightly off-topic, but instead of what this list is showing now, I think it should actually show a list of all the installed mods, the mod name and the version details need not to be retrieved from the mod files, but instead, should be fetched and stored locally when the mods are first downloaded & installed as decribed above.

I am not sure if you are familiar with MO2, but coming from Skyrim and Fallout4 modding, I think it also will be easier to manage you use symlink to install the mods instead of how you currently do it. Install the mods into the app's directory, symlink into SPT folders. Need to remove the mods? Need to disable some mods for testing? Just remove the symlinks. Reduces the need to do file I/O as well.

**Reply by SchraderR** [reply]
Hey @maxloo2,

thanks for your feedback. I will try to cover all points.

#### _SPT Hub ID to their mod is more or less impossible without the help from SPT staffs_
I think the SPT Staff knows about this topic and there was a similar message about this (from the staff). Even after releasing an API, it's the same issue without having a link between installed mods and mods hosted on the website.
Even when it's not enforced, it would be only a small change but will speed run the update process. If the mod authors don't want to add it, there will be no support for the update process (currently, maybe different with a better workflow). 

#### _the update flow: mod list has to be saved locally_
My plan was to update something similar into the config repo, https://github.com/SchraderR/sp-tarkov-client-configuration. Add a new array and populate it with mod data. Only the most important things.

The main problem is, I was to dump to implement the flow that will retrieve the mod name at the end of the installation.
So I implement a mapping list with the names of the mods (from the assembly or package.json) and how it's named at the hub.
Names will change and URL too, that's why ID would be the best because it would get the info of mods the user installed without the manager.

The current update process would need the mapping of local mod name and the name used at the hub, after this It's fetching the current available versions and compare everything / notify the user.

##### some are split between client & server
Yea. I would need a service to save installed files of mods so it can be reverted / disabled or similar for the whole mod and not just client or server part.
(Or use the hubId, and you could do all actions for the client/server mod with the correct ID without effort)

#### _Slightly off-topic_
I would disagree and like the fact it's working with mods installed outside the app, so it's not restricted to be used. Saving after installation would be conflicting when the user first installed a mod with the app and installed later a new version without the manager. It would create some conflicts.

#### _symlink_
Read something similar and your points are quite good, but I would like to implement this later. Maybe after we implemented most basic features. But sounds good!


I hope I didn't miss anything and could answer some questions :)

**Reply by SchraderR** [reply]
> ![image](https://private-user-images.githubusercontent.com/4709356/332533245-bde83d13-c507-4a16-bf50-b6993110b94a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcxNTQ0NTEsIm5iZiI6MTcyNzE1NDE1MSwicGF0aCI6Ii80NzA5MzU2LzMzMjUzMzI0NS1iZGU4M2QxMy1jNTA3LTRhMTYtYmY1MC1iNjk5MzExMGI5NGEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MjRUMDUwMjMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmIzOTM5Nzc3ZTNiNDlmMzA4ZWMxZTZkYzgyMTk1MTkzNjU5YzhlNzk1NzI3MzdhODdmYmQzZTA1ZWI2ZTYyNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.kvyGsbSdMJii6boB2UTE6vgsrj2mFe4TtcSFtDN0ZwY) Slightly off-topic, but instead of what this list is showing now, I think it should actually show a list of all the installed mods, the mod name and the version details need not to be retrieved from the mod files, but instead, should be fetched and stored locally when the mods are first downloaded & installed as decribed above.
> 
> I am not sure if you are familiar with MO2, but coming from Skyrim and Fallout4 modding, I think it also will be easier to manage you use symlink to install the mods instead of how you currently do it. Install the mods into the app's directory, symlink into SPT folders. Need to remove the mods? Need to disable some mods for testing? Just remove the symlinks. Reduces the need to do file I/O as well.

@maxloo2 In the end, i will do it like you described. Its not managable / too much work to keep up with changes. This will make it better for future features or similar.

**Comment by SchraderR** [comment]
Current workflow will only support the mods listened in the indexed mod list. Getting the hub ID works okay but fetching the newest version for all mods through the hub will generate a lot of load and the current performance (of the hub / loading times) is not good at all.

We need to wait for the official API to support all mods.
